BSP connection (WIP) [DO NOT MERGE]

tudny opened the pull request:
<!--
Hello and thank you for the pull request!

We don't have any strict rules about pull requests, but you might check
https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md
for some hints!

Also, please write a short description explaining your change in the following format: `changelog: %description%`
This description will help a lot to create release changelog. 
Drop these lines for internal only changes

:)
-->

changelog:

Kobzol committed:
INT: generalize MatchToIfLetIntention to arbitrary arm count

tudny commented:
Connected thanks to @Xewar313 and debugging session.

abrams27 reviewed (commented):
also the connection service imo is a bit messy, but it's still in progress iianm

Kobzol committed:
IDE: add implementations Code Vision hint

mchernyavsky committed:
RUN: Support `Emulate terminal in output console` for Debug runner

abrams27 commented on src/main/kotlin/org/rust/bsp/service/BspConnectionServiceImpl.kt:
```diff
@@ -0,0 +1,229 @@
+
+    override fun connect() {
+        println("Starting BSP server")
+        getBspServer()
```
does it mean that `connect()` == `getBspServer()`?

vlad20012 committed:
T: add `RustCompilerErrorComparisonTest`

This test runs our inspections against the Rust compiler's test suite
and reports detected false negatives and false positives.
You can find a verbose test report in `build/reports` directory.

tudny commented on src/main/kotlin/org/rust/cargo/toolchain/tools/Cargo.kt:
```diff
@@ -170,8 +169,8 @@ class Cargo(
-        val buildScriptsInfo = if (isFeatureEnabled(RsExperiments.EVALUATE_BUILD_SCRIPTS) && !useBSP) {
-            val listener = listenerProvider(CargoCallType.BUILD_SCRIPT_CHECK)
+        val buildScriptsInfo = if (isFeatureEnabled(RsExperiments.EVALUATE_BUILD_SCRIPTS)) {
+                val listener = listenerProvider(CargoCallType.BUILD_SCRIPT_CHECK)
```
Whitespace mistype?

abrams27 commented on src/main/kotlin/org/rust/bsp/BspClient.kt:
```diff
@@ -0,0 +1,48 @@
+
+    override fun onBuildShowMessage(params: ShowMessageParams?) {
+        println("onBuildShowMessage $params")
+        log.info("onBuildShowMessage $params")
```
bump

abrams27 commented on src/main/kotlin/org/rust/bsp/BspConstants.kt:
```diff
@@ -0,0 +1,14 @@
+    // TODO: This should be removed, as not all BSP projects are Bazel projects
+    const val BSP_WORKSPACE = "WORKSPACE"
+
+    //TODO make sure that this is always true
```
it's kinda true, but we are about to remove this mechanizm anyway xd

abrams27 commented on src/main/kotlin/org/rust/bsp/BspStartupActivity.kt:
```diff
@@ -0,0 +1,31 @@
+ * found in the LICENSE file.
+ */
+
+class BspStartupActivity : StartupActivity.DumbAware {
```
`StartupActivity` is deprecated, u should use `ProjectActivity`

abrams27 commented on src/main/kotlin/org/rust/bsp/BspStartupActivity.kt:
```diff
@@ -0,0 +1,31 @@
+class BspStartupActivity : StartupActivity.DumbAware {
+
+    override fun runActivity(project: Project) {
+        doRunActivity(project)
```
u should check here is it a rust (bsp) project - otherwise this thing will be executed even if someone opens a gradle project

abrams27 commented on src/main/kotlin/org/rust/cargo/toolchain/impl/CargoMetadata.kt:
```diff
@@ -5,6 +5,7 @@
 
 package org.rust.cargo.toolchain.impl
 
+import ch.epfl.scala.bsp4j.*
```
is it needed?

abrams27 commented on src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildManager.kt:
```diff
@@ -102,6 +106,19 @@ object CargoBuildManager {
         }
 
+        var usesBsp = cargoProject.workspace?.usesBSP
+        if (usesBsp != null && usesBsp) {
```
u can just compare it with true https://www.baeldung.com/kotlin/boolean-in-if-statement#comparing-with-a-value

abrams27 commented on src/main/kotlin/org/rust/bsp/service/BspConnectionServiceImpl.kt:
```diff
@@ -0,0 +1,507 @@
+    // Otherwise, it's just a normal symlink.
+    val normalisedWorkspace = projectDirectory.normalize().toString() + "/"
+    val replacer: (String) -> String = replacer@{
+        // TODO: there must be a better way to do this
```
cant u use `URI` here?

Commit:
ANN: Change bare trait object lint to error E0782

konrad-sz committed:
INSP: Enable path statements inspection by default

INSP: Use needsDrop in path statement inspections | Add statement drops value warning and fix for it

INSP: Implement needsDrop for simple cases

vlad20012 committed:
MIR: add deref and borrow adjustments

vlad20012 committed:
T&MIR: properly render function calls in tests

mchernyavsky committed:
MIR: Support range expressions

dima74 committed:
MIR: Add `MirProjectionElem.Downcast`

vlad20012 committed:
MIR: add method calls

mchernyavsky committed:
MIR: Implemented `if-let` expression

mchernyavsky committed:
MIR: Implemented `while-let` expression

dima74 committed:
MIR: Add `MirRvalue.Discriminant` and other subclasses

dima74 committed:
Refactor `MirPrettyPrinter.format` for `MirTerminator`

And support more than two targets

dima74 committed:
MIR: Support match for fieldless enum variants

vlad20012 committed:
Cleanup

vlad20012 committed:
MIR: add a failing test for overloaded deref

mchernyavsky committed:
Fix Scope tests

mchernyavsky committed:
Fill variants in `exprToPlace`

dima74 committed:
MIR: Support match for enum variants with fields

mchernyavsky committed:
Add Advanced Settings usages collector statistic

(cherry picked from commit a0430101f305a40ace0cc24dda1fae7759fb294a)

mchernyavsky committed:
Get rid of `CargoTestToolWindowUsagesCollector`

(cherry picked from commit 7bc29c482f4a6877999ab26b587ce8c3474e8c1d)

LebedevEA committed:
MIR: Add int to int cast

dima74 committed:
MIR: Support wildcard pattern

dima74 committed:
MIR: Support RsPatIdent for enum variant

dima74 committed:
MIR: Support RsPatTup

dima74 committed:
MIR: Support RsPatStruct

dima74 committed:
MIR: Support RsPatRef

dima74 committed:
MIR: Adjust comment for MirBuildTest

dima74 committed:
MIR: Add links to rustc source code

vlad20012 committed:
Move `fetchRustcInfo` into `CargoSyncTask` class

vlad20012 committed:
Remove timeout and add progress when fetching rustc version

This fixes the "Failed to get project sysroot" error
 when opening a project with `rust-toolchain.toml` file
 the first time and the toolchain is not installed

dima74 committed:
MIR: Support path expr named const

[skip ci]

mchernyavsky committed:
COMP: Provide postfix completion items '.sublist' and '.slice' on indexable expressions that will complete into `[]`

bors[bot] committed:
Merge #10623

10623: Next bunch of MIR r=vlad20012 a=vlad20012



Co-authored-by: vlad20012 <beskvlad@gmail.com>
Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>
Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>
Co-authored-by: Egor Lebedev <egor.lebedev@jetbrains.com>

dima74 committed:
Complete enum variants inside patterns

[skip ci]

mchernyavsky committed:
ANN: Add annotation and QF for E0557

bors[bot] committed:
Merge #10628

10628: Remove timeout and add progress when fetching rustc version  r=vlad20012 a=vlad20012

This fixes the "Failed to get project sysroot" error when opening a project with `rust-toolchain.toml` file the first time and the toolchain is not installed

changelog: Fix `Failed to get project sysroot` error when opening a project with `rust-toolchain.toml` file the first time and the toolchain is not installed

Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10629

10629: MIR: Support path expr named const r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

bors[bot] committed:
Merge #10630

10630: ANN: Add annotation and QF for E0557 r=mchernyavsky a=mchernyavsky

Relates to https://github.com/intellij-rust/intellij-rust/issues/886.

changelog: Add annotation and quick-fix for [E0557](https://doc.rust-lang.org/error_codes/E0557.html)


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

bors[bot] committed:
Merge #10631

10631: COMP: Provide postfix completion items `.sublist` and `.slice` on indexable expressions that will complete into `[]` r=mchernyavsky a=mchernyavsky

changelog: Provide postfix completion items `.sublist` and `.slice` on indexable expressions that will complete into `[]`


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

missingdays committed:
COMP: Add `as` keyword completion

[skip ci]

bors[bot] committed:
Merge #10634

10634: COMP: Add `as` keyword completion r=missingdays a=missingdays

changelog: Add `as` keyword completion


Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

Undin committed:
GDB: initial support for GDB debugging outside CLion

Undin committed:
GRD: update 2023.2 IDE and plugin dependencies

missingdays committed:
COMP: Suggest keys inside TOML dependency

[skip ci]

bors[bot] committed:
Merge #10635

10635: COMP: Suggest keys inside TOML dependency r=missingdays a=missingdays

changelog: Suggest keys inside TOML dependency


Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

Undin committed:
GDB: avoid `Error while executing Python code`

[skip ci]

bors[bot] committed:
Merge #10636

10636: Complete enum variants inside patterns r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

mchernyavsky committed:
GRAM&ANN: Permissive parsing for default parameter values

(cherry picked from commit 4fc293748f14a3edc73d6bd5fa918d887702c792)

missingdays committed:
COMP: Add completion for literal suffix

[skip ci]

bors[bot] committed:
Merge #10637

10637: COMP: Add completion for literal suffix r=missingdays a=missingdays

changelog:  Add completion for literal suffixes


Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

dima74 committed:
Fix deprecations in RsContextType

dima74 committed:
Introduce `Expression` context for live templates

[skip ci]

bors[bot] committed:
Merge #10640 #10641

10640: GRAM&ANN: Permissive parsing for default parameter values r=mchernyavsky a=mchernyavsky

changelog: Permissive parsing for default parameter values


10641: Add Advanced Settings usages collector statistics r=mchernyavsky a=mchernyavsky

changelog: Add `Advanced Settings` usages collector statistics


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

bors[bot] committed:
Merge #10119

10119: RUN: Support `Emulate terminal in output console` for Debug runner r=mchernyavsky a=mchernyavsky

Fixes https://github.com/intellij-rust/intellij-rust/issues/9778.

Relates to https://github.com/intellij-rust/intellij-rust/issues/5070.

changelog: Support `Emulate terminal in output console` for Debug runner on 2023.1


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

intellij-rust-bot committed:
Update rustc and clippy lints

intellij-rust-bot committed:
:arrow_up: patch version

bors[bot] committed:
Merge #10643

10643: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

bors[bot] committed:
Merge #10645

10645: Fix live templates suggestion in loop labels and other positions r=dima74 a=dima74

Currently we have `RsContextType.Statement` which accepts wide range of contexts, e.g. those locations:
```
let x = /*caret*/
func(/*caret*/);
break '/*caret*/
```
Thus live templates such as `assert!();` are incorrectly suggested in those locations. This PR changes `RsContextType.Statement` to accept only top-level statements inside RsBlock, and introduces `RsContextType.Expression` which equals to current `RsContextType.Statement`.

Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

vlad20012 committed:
Show `Unresolved Reference` error for type-independent paths

Undin committed:
COMP: complete Cargo instructions in build scripts

[skip ci]

bors[bot] committed:
Merge #10647

10647: COMP: complete Cargo instructions in build scripts r=mchernyavsky a=Undin

![cargo-instruction-completion](https://github.com/intellij-rust/intellij-rust/assets/2539310/f58d909d-ac64-4050-9bed-aecd782fb7e4)


changelog: Complete Cargo [instructions](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script) in build scripts


Co-authored-by: Arseniy Pendryak <arseniy.pendryak@jetbrains.com>

wbars committed:
New inspection: invalid keyword in cargo.toml

mchernyavsky committed:
INSP: Fix false-positive `Use of moved value` for lambda expressions in `for` loops

mchernyavsky committed:
Cleanup

bors[bot] committed:
Merge #10428

10428: (INT): Fix `Add unsafe to function` quickfix false positives r=mchernyavsky a=DeoTimeTheGithubUser

changelog:
 - The `add unsafe to function` quickfix is no longer raised when an unsafe call is made inside a test function, doctest main function, or in a implemented function where the parent trait function is not marked as unsafe. Suggests `Surround with unsafe block` instead.
 - Fixes #10312
 - Fixes #10183

Co-authored-by: deotime <89555032+DeoTimeTheGithubUser@users.noreply.github.com>

mchernyavsky committed:
ANN: Use E0403 instead of E0263

(cherry picked from commit 3b5607865bd3650eb39aab7a58f23de644e1611b)

mchernyavsky committed:
COMP: Show the value of the constant in the completion list

(cherry picked from commit a40101afcca0aeba6788bab9bca7724dd7bd65c8)

mchernyavsky committed:
COMP: Add `iter` alias for `for` live template

(cherry picked from commit a15ca955310f7ec56bc0d08f2a6978ba9dc59863)

bors[bot] committed:
Merge #10648

10648: Show `Unresolved Reference` error for type-independent paths r=vlad20012 a=vlad20012

After enabling attribute macro expansion I think it's time to enable `Unresolved Reference` inspection in the case of a type-independent path.


What should be highlighted:

1. Unresolved reference in a `use`:

```rust
use foo:bar;
```

2. Unresolved 1-segment path:

```rust
fn main() {
    let b = a + 1;
}
```

3. Unresolved multi-segment path if the penultimate segment is a module:

```rust
mod foo {}
fn main () {
    let a = foo::bar;
}
```

4. Unresolved multi-segment path if the penultimate segment is a enum and the last segment starts with an uppercase letter:

```rust
enum Foo { A, B }
fn main() {
    let a = Foo::C;
}
```

What should NOT be highlighted:

1. A non-path, e.g. a method call:

```rust
struct S;
fn foo() {
    S.unresolved(); // NOT highlighted
}
```

2. A multi-segment path if the penultimate segment is not a module

```rust
struct S;
fn foo() {
    let a = S::foo; // NOT highlighted
}
```

3. An explicit type-qualified path

```rust
struct S;
fn foo() {
    let a = <S>::foo; // NOT highlighted
    let a = <S as Default>::foo; // NOT highlighted
}
```


changelog: Show `Unresolved Reference` error for type-independent paths


Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10649

10649: New inspection: invalid keyword in cargo.toml r=wbars a=wbars



Co-authored-by: Kirill Smelov <kirill.smelov@jetbrains.com>

missingdays committed:
Don't complete version values for a key that is not `version`

bors[bot] committed:
Merge #10436

10436: (ANN): Change bare trait object lint to error E0782 r=mchernyavsky a=DeoTimeTheGithubUser

Trait objects without the `dyn` keyword became an error ([E0782](https://doc.rust-lang.org/error_codes/E0782.html)) in the 2021 Rust edition, however the inspection for it only raised a warning that it was deprecated.

changelog:
 - The "trait objects without dyn keyword" inspection raises an error in the 2021 edition, and a warning on 2018.

Co-authored-by: BuildTools <deotimethegithubuser@users.noreply.github.com>

bors[bot] committed:
Merge #10654

10654: INSP: Fix false-positive `Use of moved value` for lambda expressions in `for` loops r=mchernyavsky a=mchernyavsky

changelog: Fix false-positive `Use of moved value` for lambda expressions in `for` loops


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

bors[bot] committed:
Merge #10650 #10651 #10653

10650: COMP: Add `iter` alias for `for` live template r=mchernyavsky a=mchernyavsky

changelog: Add `iter` alias for `for` live template


10651: COMP: Show the value of the constant in the completion list r=mchernyavsky a=mchernyavsky

changelog: Show the value of the constant in the completion list



10653: ANN: Use E0403 instead of deprecated E0263 r=mchernyavsky a=mchernyavsky

changelog: Use [E0403](https://doc.rust-lang.org/error_codes/E0403.html) instead of deprecated [E0263](https://doc.rust-lang.org/error_codes/E0263.html)


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

vlad20012 committed:
Enable `Import` quick-fix inside an attribute macro

yopox committed:
INSP: Add a TOML inspection for duplicated keys (in simple cases)

bors[bot] committed:
Merge #10656

10656: COMP: Don't complete version values for a key that is not `version` r=vlad20012 a=vlad20012



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

mchernyavsky committed:
PARSE&ANN: Permissive parsing of struct inheritance

mchernyavsky committed:
GRAM&ANN: Permissive parsing for `impl` and `dyn` in type parameters bounds

dima74 committed:
Don't add new line by smart enter action after `;` completion

[skip ci]

konrad-sz committed:
INT: Add quick fix for E0185 and E0186

bors[bot] committed:
Merge #10657

10657: Enable `Import` quick-fix inside an attribute macro r=vlad20012 a=vlad20012



Co-authored-by: vlad20012 <beskvlad@gmail.com>

wbars committed:
Cleanup

wbars committed:
i18n: Extract hardcoded strings into RsBundle, enable i18n inspections

bors[bot] committed:
Merge #10659

10659: INT: Add quick fix for E0185 and E0186 r=konrad-sz a=konrad-sz

changelog: Add quick fix for [E0185](https://doc.rust-lang.org/error_codes/E0185.html) and [E0186](https://doc.rust-lang.org/error_codes/E0186.html)


Co-authored-by: konrad.sztenderski <konrad.sztenderski@jetbrains.com>

intellij-rust-bot committed:
Update rustc and clippy lints

bors[bot] committed:
Merge #10592

10592: INSP: Enable path statements inspection by default r=konrad-sz a=konrad-sz

changelog:
Add better support for path statements inspection and enable it by default

Co-authored-by: konrad.sztenderski <konrad.sztenderski@jetbrains.com>

bors[bot] committed:
Merge #10661

10661: GRAM&ANN: Permissive parsing for `impl` and `dyn` in type parameters bounds r=mchernyavsky a=mchernyavsky

changelog: Permissive parsing for `impl` and `dyn` in type parameters bounds


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

wbars committed:
i18n: Cleanup Shorten some keys

vlad20012 committed:
Update change notes link

bors[bot] committed:
Merge #10660

10660: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

ceronman committed:
INSP: main function is async (E0752)

https://doc.rust-lang.org/error_codes/E0752.html

Additionally added a simple quick fix to remove the async modifier
from the function.

bors[bot] committed:
Merge #10663

10663: Don't add new line by smart enter action after `;` completion r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

bors[bot] committed:
Merge #10664

10664: INSP: main function is async (E0752) r=ceronman a=ceronman

https://doc.rust-lang.org/error_codes/E0752.html

Additionally added a simple quick fix to remove the async modifier from the function.

changelog: Added new inspection for when entry point is async E0752.

Co-authored-by: Manuel Ceron <ceronman@gmail.com>

Undin committed:
INSP: use new API from to detect grammar errors

Should also fix #7446 and #7680

[skip ci]

mryall-mawson committed:
#10125: support multiple build targets in .cargo/config.toml

- add tests for parsing multiple targets
- change `buildTarget: String?` to `buildTargets: List<String>` in Config object and related function calls
- add test for multiple custom build targets (JSON files)
- use empty/singleton list as relevant defaults instead of null/host triple

bors[bot] committed:
Merge #10666

10666: INSP: Add a TOML inspection for duplicated keys (in simple cases) r=vlad20012 a=yopox

changelog: Highlight duplicated keys from the same table in TOML files.


Co-authored-by: Louis Vignier <louis.vignier@jetbrains.com>

bors[bot] committed:
Merge #10668

10668: i18n: Extract hardcoded strings into RsBundle, enable i18n inspections r=vlad20012 a=vlad20012

changelog: Extract hardcoded strings into a message bundle, allowing its future internationalization


Co-authored-by: Kirill Smelov <kirill.smelov@jetbrains.com>

missingdays committed:
Normalize parameter values after substitution in struct pattern matching

[skip ci]

vlad20012 committed:
MACRO: fix error message for non-expanded macro inside an impl

[skip ci]

vlad20012 committed:
RES: Fix name resolution inside a nested block under cfg attribute

[skip ci]

vlad20012 committed:
MACRO: Increase proc macro expansion timeout to 20 seconds

bors[bot] committed:
Merge #10669

10669: Normalize parameter values after substitution in struct pattern matching r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

vlad20012 committed:
i18n: move and messages related to macro expansion errors

vlad20012 committed:
MACRO: Improve error message for too large macro expansion

[skip ci]

bors[bot] committed:
Merge #10670

10670: GDB: Initial support for gdb debugging outside CLion r=mchernyavsky a=Undin

- the feature is still under `org.rust.debugger.gdb.setup.v2` registry option
- works only since 232.8296
- works only on Linux local debugging. Windows doesn't work because of `winbreak` binaries are expected to be present in Native Debugging plugin but it doesn't contain them. I think it will be fixed on Native Debugging plugin side and won't require any changes from the Rust plugin
- WSL debugging is not supported

changelog: Implement initial support for gdb debugging outside of CLion. Right now it's disabled by default and works only on Linux


Co-authored-by: Arseniy Pendryak <arseniy.pendryak@jetbrains.com>

bors[bot] committed:
Merge #10671

10671: INSP: use new API from to detect grammar errors r=vlad20012 a=Undin

These changes:
- drop usage of old `GrammarCheckingStrategy` API in favor of new `TextExtractor`
- enable the corresponding tests which were disabled quite a while ago
- fix #7446 and #7680

changelog: Fix bugs related to grammar error detection in Rust code using new API from [Grazie](https://plugins.jetbrains.com/plugin/12175-grazie-lite) plugin. Note, grammar error detection is disabled by default for now. You can enable it for Rust in `Scope` tab of `Preferences | Editor | Natural Languages | Grammar and Style` settings


Co-authored-by: Arseniy Pendryak <arseniy.pendryak@jetbrains.com>

Commit:
INT: Adding a quick fix to add #[tokio::main]

missingdays committed:
Fix dependency keys completion when another key exists before

[skip ci]

bors[bot] committed:
Merge #10675

10675: MACRO: fix error message for non-expanded macro inside an impl r=vlad20012 a=vlad20012

changelog: Improve error message for a non-expanded macro inside an impl block

Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10676

10676: RES: Fix name resolution inside a nested block under cfg attribute r=vlad20012 a=vlad20012


Example:
```rust
fn main() {
    {
        let mut a = 1;
              //X
        #[cfg(intellij_rust)]
        {
            use foo::bar;
            a += 1;
        } //^
        a
    };
}
```

changelog: Fix name resolution inside a nested block under cfg attribute

Co-authored-by: vlad20012 <beskvlad@gmail.com>

vlad20012 committed:
Fix stdlib deps fetching on 1.72.0-beta - use RUSTC_BOOTSTRAP

bors[bot] committed:
Merge #10677 #10678

10677: INT: Adding a quick fix to add #[tokio::main] r=ceronman a=ceronman

When `async` main function is found, this will suggest to use `#[tokio::main]`.
If `tokio` is not a dependency of the project, it will be added via `cargo add`
This will also suggest to use `#[tokio::main]` if `await` is used inside a non-async main.

changelog: Add quick-fix to install tokio and add #[tokio::main]


10678: MACRO: Increase proc macro expansion timeout to 20 seconds, improve error message in the case of a large expansion r=vlad20012 a=vlad20012

Part of #10624

1. Increase proc macro expansion timeout to 20 seconds
2. Improve error message in the case of too large macro expansion

changelog: Increase proc macro expansion timeout to 20 seconds, Improve error message in the case of too large macro expansion

Co-authored-by: Manuel Ceron <manuel.ceron@jetbrains.com>
Co-authored-by: vlad20012 <beskvlad@gmail.com>

dima74 committed:
Add completion for `fn main() { ... }`

[skip ci]

dima74 committed:
Fix scope for `macro` template

[skip ci]

vlad20012 committed:
T: Check intention/quick fix preview after checking intention invocation

Previously we checked intention/quick fix preview *before* checking
 a normal intention behavior (without preview) which was very unhandy
 in a TDD workflow. As a developer, first I want to make intention work,
 and only after that I should care about preview.

Now we check intention/quick invocation first, and only after its
 success we check the preview.

Commit:
Fixed bug when #[tokio::main] is added to a non-async main.

Also added a few more tests.

vlad20012 committed:
MACRO: match negative literals with `$l:literal` macro fragment spec

mchernyavsky committed:
GRAM&TY&ANN: Permissive parsing for prefix/postfix operators

mchernyavsky committed:
Get rid of deprecated `AnnotationHolderImpl`

mchernyavsky committed:
PARSE&TY&ANN: Support `c_str_literal` feature

Undin committed:
GRD: make 2023.2 the default platform for development

bors[bot] committed:
Merge #10674

10674: Fix dependency keys completion when another key exists before r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

missingdays committed:
COMP: Show completion auto-popup when literal suffix is typed

[skip ci]

bors[bot] committed:
Merge #10679

10679: COMP: Show completion auto-popup when literal suffix is typed r=missingdays a=missingdays

changelog: Show completion auto-popup when literal suffix is typed


Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

missingdays committed:
COMP: Only complete float suffixes for literals in scientific notation

[skip ci]

bors[bot] committed:
Merge #10680

10680: COMP: Only complete float suffixes for literals in scientific notation r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

wbars committed:
Convert "Remove dbg!()" intention to inspection with INFORMATION level

bors[bot] committed:
Merge #10681

10681: PARSE&ANN: Permissive parsing of struct inheritance r=mchernyavsky a=mchernyavsky

changelog: Improve the error message when a user tries to inherit one struct from another (like `struct Foo: Bar {}`) which is not supported in Rust


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

vlad20012 committed:
Replace `RsTypeAliasIndex` with `RsAliasIndex`

bors[bot] committed:
Merge #10682

10682: Fix scope for `macro` template r=dima74 a=dima74

changelog: Complete `macro` live template only in item context

Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

bors[bot] committed:
Merge #10684

10684: GRAM&TY&ANN: Permissive parsing for prefix/postfix operators r=mchernyavsky a=mchernyavsky

changelog: Permissive parsing for prefix/postfix operators


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

missingdays committed:
Fix recently introduced i18n problems

[skip ci]

bors[bot] committed:
Merge #10686

10686: Fix recently introduced i18n problems r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

vlad20012 committed:
GRD: merge `toml` module source into the main module

dima74 committed:
Complete `fn` in type context

[skip ci]

dima74 committed:
Complete chained iterator methods

[skip ci]

bors[bot] committed:
Merge #10689

10689: Complete chained iterator methods r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

wbars committed:
Merge branch 'master' into smelov/dbg_inspection

vlad20012 committed:
Add `org.rust.mir.move-analysis` experimental feature

Now it is separated from `org.rust.mir.borrow-check`

bors[bot] committed:
Merge #10687 #10688

10687: Convert "Remove dbg!()" intention to inspection with INFORMATION level r=wbars a=wbars



10688: MACRO: match negative literals with `$l:literal` macro fragment spec r=vlad20012 a=vlad20012

This also fixes `i32` methods name resolution and completion.

changelog: match negative numeric literals with `$l:literal` macro fragment spec


Co-authored-by: Kirill Smelov <kirill.smelov@jetbrains.com>
Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10690 #10691 #10692

10690: T: Check intention/quick fix preview after checking intention invocation r=vlad20012 a=vlad20012

Previously we checked intention/quick fix preview *before* checking
 a normal intention behavior (without preview) which was very unhandy
 in a TDD workflow. As a developer, first I want to make intention work,
 and only after that I should care about preview.

Now we check intention/quick invocation first, and only after its
 success we check the preview.


10691: Add completion for `fn main() { ... }` r=dima74 a=dima74



10692: Complete `fn` in type context r=dima74 a=dima74



Co-authored-by: vlad20012 <beskvlad@gmail.com>
Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

bors[bot] committed:
Merge #10693

10693: Add `org.rust.mir.move-analysis` experimental feature r=vlad20012 a=vlad20012

Now it is separated from `org.rust.mir.borrow-check`


Co-authored-by: vlad20012 <beskvlad@gmail.com>

intellij-rust-bot committed:
Update rustc and clippy lints

intellij-rust-bot committed:
:arrow_up: stable

intellij-rust-bot committed:
:arrow_up: patch version

bors[bot] committed:
Merge #10701

10701: Fixed bug when #[tokio::main] is added to a non-async main. r=ceronman a=ceronman

#10677 introduced a bug when `#[tokio::main]` was added without async.
Also added a few more tests.

changelog: Fixed bug when #[tokio::main] is added to a non-async main.


Co-authored-by: Manuel Ceron <manuel.ceron@jetbrains.com>

missingdays committed:
COMP: Don't show literal suffix completion if the literal already contains a suffix

[skip ci]

bors[bot] committed:
Merge #10702

10702: COMP: Don't show literal suffix completion if the literal already contains a suffix r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

wbars committed:
i18n for the experimental features

bors[bot] committed:
Merge #10698 #10703

10698: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



10703: i18n for the experimental features r=wbars a=wbars



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>
Co-authored-by: Kirill Smelov <kirill.smelov@jetbrains.com>

Undin committed:
Always run tests in headless mode

Previously, we relied on that `com.intellij.testFramework.UITestUtil.getAndSetHeadlessProperty` set up headless mode for test environment. But starting from 2023.2 the platform has some code (for example, `com.intellij.ui.scale.JBUIScale.computeSystemFontData`) which is supposed to be called at the very early initialization stage, that's why it uses `java.awt.GraphicsEnvironment.isHeadless` instead.
At the same time, `java.awt.GraphicsEnvironment.isHeadless` reads `java.awt.headless` only once and remembers the result, and it is initialized in tests before `UITestUtil.getAndSetHeadlessProperty` call. So, if `java.awt.headless` is not specified, some default value will be used and stored. For example, it might be false even on CI.
As a result, `com.intellij.ui.scale.JBUIScale.computeSystemFontData` considered that we are run with some graphics environment and failed because some contracts were not satisfied.

These changes provide `java.awt.headless=true` system property for all tests to instruct Swing to use headless mode in any cases and avoid unexpected issues with code which use `java.awt.GraphicsEnvironment.isHeadless` directly

[skip ci]

mchernyavsky committed:
Update change notes link

Commit:
INSP: New inspection to verify values out of range.

bors[bot] committed:
Merge #10705

10705: GRD: merge `toml` module source into the main module r=vlad20012 a=vlad20012

changelog: Merge `toml` module source into the main module


Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10706

10706: Fix stdlib deps fetching on 1.72.0-beta r=vlad20012 a=vlad20012

Starting from `1.72.0-beta.1` there is a nightly Cargo feature usage in the stdlib:
`cargo-features = ["public-dependency"]`
So we need to pass `RUSTC_BOOTSTRAP=1` during loading the stdlib workspace.

changelog: Fix stdlib deps fetching on 1.72.0-beta

Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10708

10708: Update 2023.2 IDE and plugin dependencies  r=Undin a=Undin

And make 2023.2 the default platform for development

changelog: Make 2023.2 the default platform for development


Co-authored-by: Arseniy Pendryak <arseniy.pendryak@jetbrains.com>

bors[bot] committed:
Merge #10699

10699: :arrow_up: stable r=Undin a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

wbars committed:
Add QF to change `a..b` to `a..=b` in match patterns

bors[bot] committed:
Merge #10710

10710: Add QF to change `a..b` to `a..=b` in match patterns r=wbars a=wbars



Co-authored-by: Kirill Smelov <kirill.smelov@jetbrains.com>

dima74 committed:
Support inspection "Unused must_use" for `impl Trait`

[skip ci]

bors[bot] committed:
Merge #10709

10709: INSP: Check if literal values are out of range. r=ceronman a=ceronman

INSP: Check if literal values are out of range.

e.g.

let x: u8 = 300

changelog: New inspection to check if integer literals are out of range


Co-authored-by: Manuel Ceron <manuel.ceron@jetbrains.com>

mchernyavsky committed:
Respect `CARGO_HOME` env variable when looking for Rust toolchain

dima74 committed:
Introduce `LookupElement.presentation` property

dima74 committed:
Improve completion of `if` and `match` keywords

Now it is similar to live template like `while`

bors[bot] committed:
Merge #10714

10714: PARSE&TY&ANN: Support `c_str_literal` feature r=mchernyavsky a=mchernyavsky

changelog: Support `c_str_literal` feature


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

bors[bot] committed:
Merge #10715 #10716

10715: Get rid of deprecated `AnnotationHolderImpl` r=mchernyavsky a=mchernyavsky



10716: Respect `CARGO_HOME` env variable when looking for Rust toolchain r=mchernyavsky a=mchernyavsky

changelog: Respect `CARGO_HOME` env variable when looking for Rust toolchain


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

mchernyavsky committed:
Fix Cargo tool window initialization on 2023.3

vlad20012 committed:
Index `use` aliases in RsAliasIndex

vlad20012 committed:
Process aliases from `RsAliasIndex` index transitively

[skip ci]

dima74 committed:
Fill match arms after `match` keyword completion

[skip ci]

LebedevEA committed:
Add QF for printing struct that without Debug or Display

[skip ci]

LebedevEA committed:
Add inspection `thread_rng().gen()` to `random()`

[skip ci]

bors[bot] committed:
Merge #10719

10719: Fix Cargo tool window initialization on 2023.3 r=mchernyavsky a=mchernyavsky

changelog: Fix Cargo tool window initialization on 2023.3


Co-authored-by: Mikhail Chernyavsky <mikhail.chernyavsky@jetbrains.com>

konrad-sz committed:
COMP: Improve completion for true/false

missingdays committed:
Consider trait bounds when performing dereference

[skip ci]

bors[bot] committed:
Merge #10721

10721: Consider trait bounds when performing dereference r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

vlad20012 committed:
RES: resolve enum variants through a type

dima74 committed:
Fix weak warnings tests without `checkWeakWarn = true`

[skip ci]

Commit:
Fixed bug causing false positives during checking literal value overflow

On one side negated values are now taken into account. Using a simple strategy of checking for unary minus, similarly to what the compiler does.

On the other side, we're not checking i32 because of type inference issues.

bors[bot] committed:
Merge #10720

10720: RES: resolve enum variants through a type r=vlad20012 a=vlad20012

changelog: resolve a enum variants through an associated type

Co-authored-by: Vladislav Beskrovny <Vladislav.Beskrovny@jetbrains.com>

bors[bot] committed:
Merge #10723

10723: T: add `RustCompilerErrorComparisonTest` r=vlad20012 a=vlad20012

This test runs our inspections against the Rust compiler's test suite and reports detected false negatives and false positives. You can find a verbose test report in `build/reports` directory.

Co-authored-by: vlad20012 <beskvlad@gmail.com>

bors[bot] committed:
Merge #10718

10718: Add macro expansion to quick doc r=vlad20012 a=LebedevEA

changelog: Add macro expansion to quick documentation on hover


Co-authored-by: Egor Lebedev <egor.lebedev@jetbrains.com>

LebedevEA committed:
Suggest associated types in completion without Self:: prefix

[skip ci]

bors[bot] committed:
Merge #10726

10726: Add inspection `thread_rng().gen()` to `random()` r=LebedevEA a=LebedevEA

changelog: Add inspection to replace `therad_rng().gen()` to `random`() from `rand` crate


Co-authored-by: Egor Lebedev <egor.lebedev@jetbrains.com>

bors[bot] committed:
Merge #10729

10729: Add quick fix for printing struct without Debug or Display r=LebedevEA a=LebedevEA

changelog: Add quick fix for printing struct without Debug or Display


Co-authored-by: Egor Lebedev <egor.lebedev@jetbrains.com>

intellij-rust-bot committed:
Update rustc and clippy lints

bors[bot] committed:
Merge #10731

10731: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

bors[bot] committed:
Merge #10734

10734: Fixed bug causing false positives during checking literal value overflow r=ceronman a=ceronman

Fixes bug introduced with #10709

On one side negated values are now taken into account. Using a simple strategy of checking for unary minus, similarly to what the compiler does.

On the other side, we're not checking i32 because of type inference issues.

changelog: Fixes bug introduced with #10709 causing negative values to be marked as overflown


Co-authored-by: Manuel Ceron <manuel.ceron@jetbrains.com>

bors[bot] committed:
Merge #10738

10738: Fix weak warnings tests without `checkWeakWarn = true` r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

bors[bot] committed:
Merge #10739

10739: Support inspection "Unused must_use" for `impl Trait` r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

neonaot committed:
CI: update dependencies in attributes-info

bors[bot] committed:
Merge #10737

10737: CI: update dependencies versions r=Undin a=neonaot

see https://github.com/rust-lang/rust/issues/113152

Fixing update-compiler-feature action, it failed with this error:
<img width="906" alt="изображение" src="https://github.com/intellij-rust/intellij-rust/assets/49211026/e7f7ebd5-c3e3-48d6-96ef-b268d5553913">
 Checked that now it is working  https://github.com/neonaot/intellij-rust/pull/90






Co-authored-by: Alina Lobanova <neonaot@gmail.com>

intellij-rust-bot committed:
Update rustc and clippy lints

intellij-rust-bot committed:
Update compiler features

Commit:
COMP: Add completion for async keyword

Commit:
Fix Completion window shows `self::`, but writes `self`

bors[bot] committed:
Merge #10741

10741: Update compiler features r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

bors[bot] committed:
Merge #10744

10744: Improve completion of `if` and `match` keywords r=dima74 a=dima74



Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

bors[bot] committed:
Merge #10742 #10743

10742: Fix Completion window shows `self::`, but writes `self` r=ceronman a=ceronman

changelog: Fix Completion window shows `self::`, but writes `self`


10743: COMP: Add completion for async keyword r=ceronman a=ceronman

changelog: Add completion for async keyword


Co-authored-by: Manuel Ceron <manuel.ceron@jetbrains.com>

Commit:
T: Added some extra tests for value out of range inspection.

bors[bot] committed:
Merge #10740

10740: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

dima74 committed:
Fix inspection "Unused must_use" for `tokio::main` proc macro

We don't expand `tokio::main` async function, so return type of such function was considered as `impl Future`, though actually `tokio::main` expands to function returning unit

[skip ci]

missingdays committed:
Add Expectation to RsTypeInferenceWalker.inferMacroAsExpr

[skip ci]

bors[bot] committed:
Merge #10748

10748: Add Expectation to RsTypeInferenceWalker.inferMacroAsExpr r=missingdays a=missingdays



Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>

missingdays committed:
Use `processElementsWithMacros` when resolving pattern binding

[skip ci]

bors[bot] committed:
Merge #10749 #10750

10749: Use `processElementsWithMacros` when resolving pattern binding r=missingdays a=missingdays



10750: Fix inspection "Unused must_use" for `tokio::main` proc macro r=dima74 a=dima74

We don't expand `tokio::main` async function, so return type of such function was considered as `impl Future`, though actually `tokio::main` expands to function returning unit

Co-authored-by: evgeny.bovykin <evgeny.bovykin@jetbrains.com>
Co-authored-by: Dmitry Murzin <dmitry.murzin@jetbrains.com>

konrad-sz committed:
COMP: Add "as" completion to provide new names

bors[bot] committed:
Merge #10746

10746: COMP: Improve completion for true/false r=konrad-sz a=konrad-sz

changelog:
Added completion for true/false in places when type is unknown

Co-authored-by: konrad.sztenderski <konrad.sztenderski@jetbrains.com>

Undin committed:
GRD: update 2023.2 IDE and plugin dependencies to release ones

[skip ci]

bors[bot] committed:
Merge #10754

10754: GRD: update 2023.2 IDE and plugin dependencies to release ones r=Undin a=Undin



Co-authored-by: Arseniy Pendryak <arseniy.pendryak@jetbrains.com>

vlad20012 committed:
Fixes add #[tokio::main] quick fix.

This changes the way we are installing tokio dependency. Instead of using `cargo add`
we're editing Cargo.toml directly. This allows us to have more control over the features.
New helper function to add cargo dependency is added as well.

vlad20012 committed:
Show VCS Code Vision hints for more items

Now an author should be shown for type alias, const/static,
 outer mod declaration (`mod foo`)
 and for a trait alias (unstable Rust feature)

vlad20012 committed:
Enable VCS Code Vision hints by default

vlad20012 committed:
Use VcsCodeVisionCurlyBracketLanguageContext

bors[bot] committed:
Merge #10755

10755: Implement `impl`s search for an alias of alias and for a use alias r=vlad20012 a=vlad20012



Co-authored-by: vlad20012 <beskvlad@gmail.com>
Co-authored-by: Vladislav Beskrovny <vladislav.beskrovny@jetbrains.com>

bors[bot] committed:
Merge #10756

10756: Enable VCS Code Vision hints by default r=vlad20012 a=vlad20012



Co-authored-by: vlad20012 <beskvlad@gmail.com>

vlad20012 committed:
Adjust intellisense file size limit for Rust files

bors[bot] committed:
Merge #10757 #10758

10757: Fixes add #[tokio::main] quick fix. r=ceronman a=ceronman

This changes the way we are installing tokio dependency. Instead of using `cargo add` we're editing Cargo.toml directly. This allows us to have more control over the features. New helper function to add cargo dependency is added as well.

changelog: Fixes undo not available after adding tokio dependency when using Add `#[tokio::main]` quickfix.


10758: T: Added some extra tests for value out of range inspection. r=ceronman a=ceronman



Co-authored-by: vlad20012 <beskvlad@gmail.com>
Co-authored-by: Manuel Ceron <manuel.ceron@jetbrains.com>

vlad20012 committed:
Enable `RsReferenceCodeVisionProvider` for top-level items

Enable it for items except:
- an associated item (an impl/trait member)
- a struct field
- a macro declaration
- an enum variant

Usage code vision can be enabled for these elements using
 `org.rust.code.vision.usage.slow` registry option

[skip ci]

bors[bot] committed:
Merge #10760

10760: Adjust intellisense file size limit for Rust files r=vlad20012 a=vlad20012

I finally figured out how to increase the file limit for Rust files. Now the limit for Rust files is 8MB. Previous attempt was #8699.

Fixes #10624

changelog: Adjust intellisense file size limit for Rust files


Co-authored-by: vlad20012 <beskvlad@gmail.com>

vlad20012 committed:
INSP: remove `unused_qualifications` lint from `unused` group

vlad20012 committed:
INSP: change `Unnecessary Qualifications` level to `WEAK WARNING`

[skip ci]

vlad20012 committed:
PERF: Optimize `RsReferenceImpl.isReferenceTo` for top level items

bors[bot] committed:
Merge #10751

10751: COMP: Add "as" completion to provide new names r=vlad20012 a=konrad-sz

changelog:
Add "as" completion to provide new names

Co-authored-by: konrad.sztenderski <konrad.sztenderski@jetbrains.com>

bors[bot] committed:
Merge #9424 #9958 #10685

9424: INT: generalize MatchToIfLetIntention to arbitrary arm count, take two r=vlad20012 a=Kobzol

This PR is another attempt of implementing https://github.com/intellij-rust/intellij-rust/pull/7435.

I learned that `if` expressions must have an `else` block if their type is anything other than `()`. So now the code always generates an `else` branch if it wasn't already generated by other means for match expressions that don't have unit type.

Fixes: https://github.com/intellij-rust/intellij-rust/issues/7409

changelog: The intention to convert a match expression to a series of if let statements is now offered in more situations.

9958: IDE: add implementations Code Vision hint r=vlad20012 a=Kobzol

The `handleClick` implementation is copied from `ImplsNavigationGutterIconRenderer::navigateToItems`. It should probably be placed into some unified location, but I'm not sure where does it belong.

Relevant issue: https://github.com/intellij-rust/intellij-rust/issues/9427

changelog: Add support for showing implementations (inheritors) Code Vision. Note, these hints are available starting with 2022.3 and are currently disabled by default. You can turn it on via `Registry... | org.rust.code.vision.implementation`.

10685: #10125: Support multiple build targets in .cargo/config.toml r=vlad20012 a=mryall-mawson

changelog: support multiple build.target values in .cargo/config.toml

This is a fix for #10125, where the `cargo metadata` execution fails when multiple build targets are specified in `.cargo/config.toml`. Multiple build targets in config.toml were stabilized in Rust 1.64.0, as part of rust-lang/cargo#8176.

As described in [my comment on the issue](https://github.com/intellij-rust/intellij-rust/issues/10125#issuecomment-1630091562), I have changed `buildTarget: String?` into `buildTargets: List<String>` in the `CargoConfig`, as returned by `Cargo.getConfig()`, and used in various places which calls that function. Instead of being `null`, `buildTargets` is an empty list if build.target is not specified, which is still correctly handled (defaulted to the host triple) by the calling code.

As a result, `Cargo.fetchMetadata()` now passes `--filter-platform` multiple times to the `cargo metadata` call, rather than an empty string, which will correctly filter the dependencies for multiple target platforms, fixing the implementation originally added in #7185.

Tests have been added and updated to exercise this functionality.

This is my first time working on this project, so please feel free to adjust whatever is needed to take this forward.


Co-authored-by: Jakub Beránek <berykubik@gmail.com>
Co-authored-by: Matt Ryall <mryall@mawsonrovers.com>

bors[bot] committed:
Merge #10730 #10762

10730: Suggest associated types in completion without Self:: prefix r=vlad20012 a=LebedevEA

changelog: Suggest associated types in completion without Self:: prefix

10762: INSP: remove `unused_qualifications` lint from `unused` group, change level to `WEAK WARNING`  r=vlad20012 a=vlad20012

changelog: remove `unused_qualifications` lint from `unused` group, change level to `WEAK WARNING` 


Co-authored-by: Egor Lebedev <egor.lebedev@jetbrains.com>
Co-authored-by: Vladislav Beskrovny <Vladislav.Beskrovny@jetbrains.com>

bors[bot] committed:
Merge #10763

10763: Enable `RsReferenceCodeVisionProvider` for top-level items r=vlad20012 a=vlad20012

The provider was initially added in #9927.

Here I enable usages Code Vision for Rust items except:
- an associated item (an impl/trait member)
- a struct field
- a macro declaration
- an enum variant

Usage code vision can be enabled for these elements using `org.rust.code.vision.usage.slow` registry option.

For other items Reference Search does not use type inference and hence it should work fast.
I additionally adjusted `RsReferenceImpl.isReferenceTo()` to ensure it does not touch type inference in such cases.

changelog:  Enable Usages Code Vision hints for some top-level items

Co-authored-by: Vladislav Beskrovny <Vladislav.Beskrovny@jetbrains.com>
Co-authored-by: vlad20012 <beskvlad@gmail.com>

intellij-rust-bot committed:
Update rustc and clippy lints

intellij-rust-bot committed:
Update compiler features

intellij-rust-bot committed:
:arrow_up: patch version

bors[bot] committed:
Merge #10764

10764: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

bors[bot] committed:
Merge #10765

10765: Update compiler features r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

LebedevEA committed:
Disable E0277 QFs for structs from deps w/o Debug & Display

LebedevEA committed:
Make "Implement `Display`" QF handle generics

Toomimi committed:
Merge with IntelliJ-Rust/master

Toomimi committed:
Try make it working

Toomimi committed:
Resolve added while merging todos

bors[bot] committed:
Merge #10790

10790: Fix bug for printing of structs w/o display & debug traits quick fixes r=LebedevEA a=LebedevEA

changelog: Fix bugs related to quick fixes of printing of structs with neither display nor debug trait implemented


Co-authored-by: Egor Lebedev <egor.lebedev@jetbrains.com>

intellij-rust-bot committed:
Update rustc and clippy lints

intellij-rust-bot committed:
Update compiler features

bors[bot] committed:
Merge #10793

10793: Update rustc and clippy lints r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

Toomimi committed:
Merge branch 'master' into merge-with-master-origin

# Conflicts:
#	src/main/kotlin/org/rust/cargo/icons/CargoIcons.kt
#	src/main/kotlin/org/rust/cargo/project/model/AttachCargoProjectAction.kt
#	src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt
#	src/main/kotlin/org/rust/cargo/project/model/impl/CargoSyncTask.kt
#	src/main/kotlin/org/rust/cargo/project/toolwindow/CargoToolWindow.kt
#	src/main/kotlin/org/rust/cargo/project/workspace/StandardLibrary.kt
#	src/main/kotlin/org/rust/cargo/toolchain/tools/Cargo.kt
#	src/main/kotlin/org/rust/ide/annotator/format/impl.kt
#	src/main/kotlin/org/rust/ide/fixes/ImplementDisplayFix.kt
#	src/main/kotlin/org/rust/lang/core/completion/lint/ClippyLints.kt
#	src/main/kotlin/org/rust/lang/core/completion/lint/RustcLints.kt
#	src/main/resources/META-INF/rust-core.xml
#	src/main/resources/compiler-info/compiler-features.json
#	src/main/resources/messages/RsBundle.properties
#	src/test/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotatorTest.kt

Toomimi committed:
Merge pull request #5 from ZPP-This-is-fine/merge-with-master-origin

Merge with master origin

bors[bot] committed:
Merge #10794

10794: Update compiler features r=neonaot a=intellij-rust-bot



Co-authored-by: intellij-rust-bot <intellij.rust@gmail.com>

Toomimi committed:
Merge branch 'intellij-rust:master' into bsp-connection

LebedevEA committed:
Add macro expansion to quick doc
